src/core/idb-cache.js

```diff
@@ -0,0 +1,109 @@
+/**
+ * A promise based interface to a key-value based indexedDB cache
+ */
+
+const databases = new Map();
+
+export default class IDBCache {
+  /**
+   * @constructor
+   * @param {string} name             name of database
+   * @param {array} stores            name of stores
+   * @param {object} Options
+   *    @param {integer} version      database version
+   *    @param {string} defaultStore  name of default store
+   */
+  constructor(name, stores = [], { version = 1, defaultStore }) {
+    this.name = name;
+    this.stores = typeof stores === "string" ? [stores] : stores;
+    this.version = version;
+    this.defaultStore = defaultStore || this.stores[0];
+    if (!databases.has(name)) {
+      const dbPromise = new Promise((resolve, reject) => {
+        const request = window.indexedDB.open(this.name, this.version);
+        request.onerror = () => reject(request.error);
+        request.onupgradeneeded = () =>
+          this.stores.forEach(storeName =>
+            request.result.createObjectStore(storeName));
+        request.onsuccess = () => resolve(request.result);
+      });
+      databases.set(this.name, dbPromise);
+    }
+  }
+
+  /**
+   * @private returns a promise to get an IDBObjectStore
+   * @param {string} storeName      store name
+   * @param {string} mode           IDBTransaction.mode
+   */
+  async ready(storeName, mode) {
+    const db = await databases.get(this.name);
+    return new Promise((resolve, reject) => {
+      const transaction = db.transaction(storeName, mode);
+      transaction.onerror = () => reject(transaction.error);
+      transaction.oncomplete = () => resolve();
```

Comment:
this line and the next compete for `resolve()`, which will be confusing.

Reply:
didn't get what you mean here.

Reply:
Sorry, will try to clarify. You have .oncomplete trying to use resolve(), but the following line calls resolve(). You can only ever call resolve once, so .oncomplete, even if it fires, won’t do anything. Make sense?